MC/Core/Src/freertos.c

```diff
@@ -0,0 +1,390 @@
+
+}
+
+/* USER CODE BEGIN Header_StartDefaultTask */
+/**
+  * @brief  Function implementing the defaultTask thread.
+  * @param  argument: Not used
+  * @retval None
+  */
+/* USER CODE END Header_StartDefaultTask */
+void StartDefaultTask(void *argument)
+{
+  /* USER CODE BEGIN StartDefaultTask */
+  /* Infinite loop */
+  for(;;)
+  {
+    osDelay(1);
+  }
+  /* USER CODE END StartDefaultTask */
+}
+
+/* Private application code --------------------------------------------------*/
+/* USER CODE BEGIN Application */
+
+void MC_ReceiveStartThread(void *argument) {
+
+	CAN_Frame_t start_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K1); //Initialize CAN_Frame to send to MC to receive first data frame
+
+	start_frame.id_type = CAN_ID_EXT;
+	start_frame.data_length = 0;
+	start_frame.rtr = 1; //Configure the request transmission bit to high
+
+	uint8_t received = 0;
+
+	for(;;){
+
+
+		// Checks if mailbox is available for transmission and sends start_frame to the MC (requesting data bacK)
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan3)) {
+			CAN_send_frame(start_frame);
+	    }
+
+		if(HAL_CAN_GetRxFifoFillLevel(&hcan3, CAN_RX_FIFO0)){
+			start_frame = CAN_get_frame(&hcan3, CAN_RX_FIFO0);
+			received = 1;
+		}
+		// Checks if message is available in RxFifo and receives that message from the MC
+
+
+		if(received){
+			osMessageQueuePut(start_frame_q, &start_frame, 0, osWaitForever);
+			osSemaphoreRelease(MCSemStartHandle);
+		}
+		//Checks if data was filled and puts the data in the queue, releasing the semaphore to indicate data is available
+
+	}
+}
+
+void MC_ReceiveEndThread(void *argument) {
+
+	CAN_Frame_t end_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K2); //Initialize CAN_Frame to send to MC to receive second data frame
+
+	end_frame.id_type = CAN_ID_EXT;
+	end_frame.data_length = 0;
+	end_frame.rtr = 1;
+
+	uint8_t received = 0;
+
+	for(;;){
+
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan3)) {
+			CAN_send_frame(end_frame);
+		}
+
+		if(HAL_CAN_GetRxFifoFillLevel(&hcan3, CAN_RX_FIFO0)){
+			end_frame = CAN_get_frame(&hcan3, CAN_RX_FIFO0);
+			received = 1;
+		}
+
+
+		if(received){
+			osMessageQueuePut(end_frame_q, &end_frame, 0, osWaitForever);
+			osSemaphoreRelease(MCSemEndHandle);
+
+		}
+
+	}
+}
+
+
+
+void MC_SendStartThread(void *argument){
+
+	CAN_Frame_t start_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K1); //Initialize CAN_frame to send to MC
+	CAN_Frame_t send_frame = CAN_frame_init(&hcan1, MOTOR_CONTROLLER_K1); //Initialize CAN_frame to send to RPI
+
+	Motor_Controller_t mc_data = {0};
+
+	for(;;){
+
+		osSemaphoreAcquire(MCSemStartHandle, osWaitForever);
+
+		osMessageQueueGet(start_frame_q, &start_frame, NULL, osWaitForever);
+
+		//Process data from MC
+		mc_data.direction = direction(CAN_get_segment(start_frame, DRIVING_DIRECTION_K));
+		mc_data.speed = CAN_get_segment(start_frame, MOTOR_SPEED_K);
+		mc_data.error_code = error_code(CAN_get_segment(start_frame, MOTOR_ERROR_CODE_K));
```

Comment:
you could theoretically do without the additional struct here if you call the processing functions within the CAN_set_segment function. I guess it helps readability though so up to you

Reply:
Leaving it for readability